### ISSUE TITLE
Handle printing of documents in browser

### ISSUE THREAD

**Issue by blink1073** [jupyterlab/jupyterlab#1314]
So you can hit the browser's print button and get a nice looking printout. This should take into account the most recent document.

**Comment by jasongrout** [comment]
Background on the print media selector - https://developer.mozilla.org/en-US/docs/Web/Guide/CSS/Getting_started/Media.

If there is a CSS class applied to the currently active document, we could maybe do something like @media print .document:not(.active-document) {display: none} or something to hide everything except the current document. Interesting...

**Comment by saulshanabrook** [comment]
I investigated the above solution and couldn't get it to work. Hiding not visible elements still leaves everything absolutely positioned by phosphor.

Then, I tried using creating an empty iframe, copying over the node we want to print into it, and calling print on the window. I used this library to do this: https://github.com/joseluisq/printd

That works for the JSON viewer, but doesn't work well for text editors or notebooks. For those, we need some custom functionality likely. For notebooks, we could use the To HTML functionality of nbconvert  to get a nicer print preview.

So then we (@afshin  @jasongrout) started talking about different ways widgets could provide custom print functionality. We settled on two ways do this that might be OK:

1. Define a new [Symbol](https://www.typescriptlang.org/docs/handbook/symbols.html) for printing and have widgets that implement custom printing define a method named by the symbol.
2. Send a message to the widget when printing.  On the `MainAreaWidget` handle this method, and call the `onPrintMessage` method which subclasses define.
